testrunner-running/src/main/scala/edu/illinois/cs/statecapture/StateCapture.java

```diff
@@ -0,0 +1,526 @@
+import com.thoughtworks.xstream.converters.reflection.FieldKeySorter;
+import com.thoughtworks.xstream.core.JVM;
+import com.thoughtworks.xstream.io.xml.DomDriver;
+
+import com.thoughtworks.xstream.mapper.Mapper;
+import com.thoughtworks.xstream.security.AnyTypePermission;
+import edu.illinois.cs.statecapture.agent.MainAgent;
+
+import java.io.File;
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.PrintWriter;
+import java.io.IOException;
+
+import java.lang.Object;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.AccessibleObject;
+import java.lang.reflect.Modifier;
+
+import java.util.List;
+import java.util.Set;
+import java.util.Map;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.TreeMap;
+import java.util.Comparator;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.nio.file.Files;
+
+import edu.illinois.cs.testrunner.configuration.Configuration;
+import edu.illinois.cs.xstream.UnmarshalChain;
+import edu.illinois.cs.xstream.CustomElementIgnoringMapper;
+import edu.illinois.cs.xstream.EnumMapConverter;
+import edu.illinois.cs.xstream.LambdaConverter;
+import edu.illinois.cs.xstream.LookAndFeelConverter;
+import edu.illinois.cs.xstream.MapConverter;
+import edu.illinois.cs.xstream.ReflectionConverter;
+import edu.illinois.cs.xstream.SerializableConverter;
+import edu.illinois.cs.xstream.TreeMapConverter;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.mockito.Mockito;
+
+import static com.thoughtworks.xstream.XStream.PRIORITY_LOW;
+import static com.thoughtworks.xstream.XStream.PRIORITY_NORMAL;
+import static com.thoughtworks.xstream.XStream.PRIORITY_VERY_LOW;
+
+public class StateCapture implements IStateCapture {
+
+
+    protected final String testName;
+    private boolean dirty;
+
+    // The State, field name of static root to object pointed to
+    private static final LinkedHashMap<String, Object> nameToInstance = new LinkedHashMap<String, Object>();
+
+    //for reflection and deserialization
+    private String subxmlFold;
+    private String rootFile;
+    private String reflectionFile;
+
+    public StateCapture(String testName) {
+        this.testName = testName;
+        setup();
+    }
+
+    static Set<String> readFileContentsAsSet(String path) {
+        File file = new File(path);
+        Set<String> keys = new HashSet<>();
+        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
+            String line;
+            while ((line = br.readLine()) != null) {
+                keys.add(line);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return keys;
+    }
+
+    private void setup() {
+        subxmlFold = Configuration.config().getProperty("statecapture.subxmlFold");
```

Comment:
Maybe we can just call this "xmlDir"; the need to distinguish "xml" and "subxml" is no longer relevant.